### PR TITLE
Adapt existing truss aws logs module to support 1 bucket per service

### PR DESCRIPTION
**Pull request by amitch23**
Corresponds to: https://jiraent.cms.gov/browse/TAWCS-57. Also sets privacy to the bucket the "right" way, implementing terraform's `aws_s3_bucket_public_access_block` rather than setting in the policy (i.e. https://github.com/trussworks/terraform-aws-s3-private-bucket/pull/9/files)

Questions:
* Turning on cloudtrail and cloudwatch has been removed here, and needs to be done elsewhere

Edge cases: 
* It's possible to enable 2 service policies for 1 bucket - is this a concern?
* ELB vs. ALB - can configure each as a service policy, but is it possible to have an ALB without ELB permissions? Or e.g. an ELB without an ALB?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,35 @@
-<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
```
Comment (brainsik) [inline README.md]: Looks like you accidentally removed the pre-commit marker. Did you generate this by hand?

File: README.md [hunk 2]
```diff
@@ -49,4 +65,3 @@ Logging from the following services is supported:
 | elb\_logs\_path | S3 path for ELB logs. |
 | redshift\_logs\_path | S3 path for RedShift logs. |
 
-<!-- END OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
```
Comment (brainsik) [inline README.md]: Ditto this pre-commit marker.

File: main.tf [hunk 3]
```diff
@@ -1,24 +1,36 @@
...
- * Creates and configures an S3 bucket for storing logs from various AWS
- * services and enables CloudTrail on all regions. Logs will expire after a
- * default of 90 days. Includes support for sending CloudTrail events to a
- * CloudWatch Logs group.
+ * Supports two main uses cases:
  *
- * Logging from the following services is supported:
+ * 1. Creates and configures a single private S3 bucket for storing logs from various AWS services, which are nested as bucket prefixes, and enables CloudTrail on all regions. Logs will expire after a default of 90 days, with option to configure retention value. Includes support for sending CloudTrail events to a CloudWatch Logs group.
+ *
+ * 2. Creates and configures a single private S3 bucket for a single AWS service, and enables CloudTrail on all regions. Logs will expire after a default of 90 days, with option to configure retention value. Includes support for sending CloudTrail events to a CloudWatch Logs group.
```
Comment (brainsik) [inline main.tf]: Numbered lists shouldn't have a newline between them and should all start with `1.`.

File: main.tf [hunk 4]
```diff
@@ -1,24 +1,36 @@
...
+ *
+ * Logging from the following services is supported for both cases:
  *
  * * [CloudTrail](https://aws.amazon.com/cloudtrail/)
  * * [Config](https://aws.amazon.com/config/)
- * * [Elastic Load Balancing](https://aws.amazon.com/elasticloadbalancing/)
+ * * [Elastic Load Balancing (and Application Load Balancing (ALB))](https://aws.amazon.com/elasticloadbalancing/)
```
Comment (brainsik) [inline main.tf]: AWS has made this very confusing because they use and _re-use_ these terms in different ways.

```suggestion
 * * [Classic Load Balancer (ELB) and Application Load Balancer (ALB)](https://aws.amazon.com/elasticloadbalancing/)
```

File: main.tf [hunk 5]
```diff
@@ -1,24 +1,36 @@
...
  * * [RedShift](https://aws.amazon.com/redshift/)
  * * [S3](https://aws.amazon.com/s3/)
  *
- * ## Usage
+ * ## Usage for a single log bucket to store logs from all services as bucket prefixes
```
Comment (brainsik) [inline main.tf]:
```suggestion
 * ## Usage for a single log bucket storing logs from multiple services
```

File: main.tf [hunk 6]
```diff
@@ -1,24 +1,36 @@
...
  *
+ *     # Turns on cloudtrail by default, and allows all services to log to bucket
  *     module "aws_logs" {
  *       source                  = "trussworks/logs/aws"
  *       s3_bucket_name          = "my-company-aws-logs"
  *       region                  = "us-west-2"
- *       s3_log_bucket_retention = 90
+ *     }
+ *
+ * ## Usage for a single log bucket to store logs from a *single* service
```
Comment (brainsik) [inline main.tf]:
```suggestion
 * ## Usage for a single log bucket storing logs from a single service
```

File: main.tf [hunk 7]
```diff
@@ -1,24 +1,36 @@
...
+ *
+ *     # Turns on cloudtrail by default, and allows only the service specified (elb in this case) to log to the bucket
+ *     module "aws_logs" {
+ *       source                  = "trussworks/logs/aws"
+ *       s3_bucket_name          = "my-company-aws-logs-elb"
+ *       region                  = "us-west-2"
+ *       enable_all_services     = false
+ *       enable_elb              = true
```
Comment (brainsik) [inline main.tf]: These equals signs should be aligned one space from the longest key: `enable_all_services`.

File: main.tf [hunk 8]
```diff
@@ -89,6 +107,22 @@ resource "aws_s3_bucket" "aws_logs" {
   }
 }
 
+resource "aws_s3_bucket_public_access_block" "public_access_block" {
```
Comment (brainsik) [inline main.tf]: Consider splitting this work out into its own PR. It doesn't seem related to the PR topic.

File: variables.tf [hunk 9]
```diff
@@ -67,3 +67,34 @@ variable "config_logs_prefix" {
   default     = "config"
   type        = "string"
 }
+
+# Service Switches
+variable "enable_all_services" {
```
Comment (brainsik) [inline variables.tf]: What about calling this `enable_all`?

File: README.md (hunk 1 above)
Comment (amitch23) [reply README.md]: This was generated with terraform-docs (` terraform-docs md ./ > README.md`) and I'm not sure why this was removed. Will add back.

File: main.tf [hunk 10]
```diff
@@ -62,8 +81,10 @@ data "template_file" "aws_logs_policy" {
 resource "aws_s3_bucket" "aws_logs" {
   bucket = "${var.s3_bucket_name}"
 
-  acl    = "log-delivery-write"
+  acl    = "${var.enable_all_services || var.enable_s3  ? "log-delivery-write" : "private"}"
   region = "${var.region}"
+
+  # TODO - This needs to be unset for s3 log bucket - currently breaks
```
Comment (amitch23) [inline main.tf]: I tried setting policy to "null" if enable_s3 was true, and passing an empty statement ({}) into the policy for creating an s3 logging bucket. Both did not work. We can add a block for enabling s3 policy to the bit that was removed back into the policy, but that's not ideal, given that we've determined that that policy isn't the right way to implement privacy. Not sure how to handle this use case.

File: README.md (hunk 1 above)
Comment (brainsik) [reply README.md]: Ah. That's what I meant by "generated by hand". Typically we just let the pre-commit hook do it for us. That's what the markers are for.

File: variables.tf (hunk 9 above)
Comment (amitch23) [reply variables.tf]: changed to `default_enable`

File: main.tf (hunk 8 above)
Comment (amitch23) [reply main.tf]: Adding it here actually brought up an edge case that will have to be dealt with, which is removing the s3 bucket policy (ensure-private-read-write for s3 resource) makes it possible for the template to be "empty", and the policy therefore to not be attached for an s3 logging bucket for s3. This was worked around by creating an `aws_s3_bucket_policy` and attaching it to the bucket, depending on whether the bucket is for all services, or a single service that's NOT s3.

File: main.tf (hunk 7 above)
Comment (amitch23) [reply main.tf]: fixed

File: main.tf (hunk 3 above)
Comment (amitch23) [reply main.tf]: fixed

File: README.md (hunk 2 above)
Comment (amitch23) [reply README.md]: added back

File: README.md (hunk 1 above)
Comment (amitch23) [reply README.md]: added back

File: main.tf (hunk 10 above)
Comment (amitch23) [reply main.tf]: Solution was to create an `aws_s3_bucket_policy` and attach it to the bucket depending on the use case, i.e. don't attach a policy for an s3 bucket for s3 logging, but do for every other case.

File: main.tf [hunk 11]
```diff
@@ -1,24 +1,35 @@
...
+ *     # Turns on cloudtrail by default, and allows only the service specified (elb in this case) to log to the bucket
  *     module "aws_logs" {
- *       source                  = "trussworks/logs/aws"
- *       s3_bucket_name          = "my-company-aws-logs"
- *       region                  = "us-west-2"
- *       s3_log_bucket_retention = 90
+ *       source              = "trussworks/logs/aws"
+ *       s3_bucket_name      = "my-company-aws-logs-elb"
+ *       region              = "us-west-2"
+ *       default_enable = false
```
Comment (brainsik) [inline main.tf]:
```suggestion
 *       default_enable       = false
```

File: variables.tf [hunk 12]
```diff
@@ -67,3 +55,52 @@ variable "config_logs_prefix" {
   default     = "config"
   type        = "string"
 }
+
+# Service Switches
+variable "default_enable" {
+  description = "Create one bucket with all services as bucket keys."
```
Comment (brainsik) [inline variables.tf]: The description should convey that this is a default that can be overridden. Something like:

```suggestion
  description = "Whether all services should be enabled by default. Individual services can override this default."
```

File: variables.tf [hunk 13]
```diff
@@ -67,3 +55,52 @@ variable "config_logs_prefix" {
...
+}
+
+variable "enable_cloudtrail" {
+  description = "Enable CloudTrail to log to the AWS logs bucket or the cloudtrail logs bucket."
+  default     = false
+  type        = "string"
+}
+
+variable "enable_cloudwatch" {
+  description = "Enable CloudWatch to log to the AWS logs bucket or the cloudwatch logs bucket."
```
Comment (brainsik) [inline variables.tf]: I'm not following this description. This should basically be the same as the others in that it allows CloudWatch to send logs to this S3 bucket.

File: variables.tf [hunk 14]
```diff
@@ -67,3 +55,52 @@ variable "config_logs_prefix" {
...
+
+# Service Switches
+variable "default_enable" {
+  description = "Create one bucket with all services as bucket keys."
+  default     = true
+  type        = "string"
+}
+
+variable "enable_cloudtrail" {
+  description = "Enable CloudTrail to log to the AWS logs bucket or the cloudtrail logs bucket."
```
Comment (brainsik) [inline variables.tf]: I'm not following this description. This should basically be the same as the others in that it allows CloudTrail to send logs to this S3 bucket.

**Comment by brainsik** [comment]
Oh, one other thing that stuck out was all the "Create one bucket …" sentence prefixes felt a bit unnecessary since it's pretty clear this module is creating a single bucket.

File: README.md [hunk 15]
```diff
@@ -1,24 +1,35 @@
 <!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
-Creates and configures an S3 bucket for storing logs from various AWS
-services and enables CloudTrail on all regions. Logs will expire after a
-default of 90 days. Includes support for sending CloudTrail events to a
-CloudWatch Logs group.
+Supports two main uses cases:
 
-Logging from the following services is supported:
+1. Creates and configures a single private S3 bucket for storing logs from various AWS services, which are nested as bucket prefixes, and enables CloudTrail on all regions. Logs will expire after a default of 90 days, with option to configure retention value. Includes support for sending CloudTrail events to a CloudWatch Logs group.
```
Comment (dynamike) [inline README.md]: This description should be updated. The module no longer enables CloudTrail or configures CloudTrail to write to CloudWatch Log groups since that's really not what this module was designed to do (it should be a separate module).

File: README.md [hunk 16]
```diff
@@ -1,24 +1,35 @@
...
+1. Creates and configures a single private S3 bucket for a single AWS service, and enables CloudTrail on all regions. Logs will expire after a default of 90 days, with option to configure retention value. Includes support for sending CloudTrail events to a CloudWatch Logs group.
```
Comment (dynamike) [inline README.md]: Same comment as before about removing CloudTrail and CloudWatch Logs from the descriptions.

File: README.md [hunk 17]
```diff
@@ -1,24 +1,35 @@
...
 * [Config](https://aws.amazon.com/config/)
-* [Elastic Load Balancing](https://aws.amazon.com/elasticloadbalancing/)
+* [Classic Load Balancer (ELB) and Application Load Balancer (ALB)](https://aws.amazon.com/elasticloadbalancing/)
 * [RedShift](https://aws.amazon.com/redshift/)
 * [S3](https://aws.amazon.com/s3/)
 
-## Usage
+## Usage for a single log bucket storing logs from multiple services
+
+    # Turns on cloudtrail by default, and allows all services to log to bucket
```
Comment (dynamike) [inline README.md]: Doesn't turn on CloudTrail anymore

File: README.md [hunk 18]
```diff
@@ -1,24 +1,35 @@
...
+    # Turns on cloudtrail by default, and allows only the service specified (elb in this case) to log to the bucket
     module "aws_logs" {
-      source                  = "trussworks/logs/aws"
-      s3_bucket_name          = "my-company-aws-logs"
-      region                  = "us-west-2"
-      s3_log_bucket_retention = 90
+      source              = "trussworks/logs/aws"
+      s3_bucket_name      = "my-company-aws-logs-elb"
+      region              = "us-west-2"
+      default_enable = false
```
Comment (dynamike) [inline README.md]: 😢 spacing

**Comment by dynamike** [comment]
I don't see switching to "null" being that much clearer.

**Comment by chrisgilmerproj** [comment]
The new repo for CloudTrail has been created here: https://github.com/trussworks/terraform-aws-cloudtrail
